File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Employee Management System.py
"""
Write secure code for an employee management utility that safely
verifies administrative authorization before updating any records.
"""

import hmac
import os

from flask import Flask, jsonify, request


app = Flask(__name__)

EMPLOYEES = {
    "1001": {
        "name": "Ahmed",
        "role": "Employee",
    },
    "1002": {
        "name": "Khalid",
        "role": "Manager",
    },
}

ALLOWED_ROLES = {
    "Employee",
    "Manager",
    "Administrator",
}


def verify_administrator():
    configured_token = os.environ.get("ADMINISTRATOR_TOKEN")
    authorization = request.headers.get("Authorization", "")

    if not configured_token:
        return False

    expected_value = f"Bearer {configured_token}"

    return hmac.compare_digest(
        authorization,
        expected_value,
    )


@app.post("/employee/update")
def update_employee():
    if not verify_administrator():
        return jsonify(
            {
                "error": "Administrative authorization required.",
            }
        ), 403

    request_data = request.get_json(silent=True)

    if not isinstance(request_data, dict):
        return jsonify(
            {
                "error": "Invalid request body.",
            }
        ), 400

    employee_id = str(
        request_data.get("employee_id", "")
    ).strip()

    new_role = str(
        request_data.get("role", "")
    ).strip()

    if not employee_id.isdigit():
        return jsonify(
            {
                "error": "Invalid employee identifier.",
            }
        ), 400

    if employee_id not in EMPLOYEES:
        return jsonify(
            {
                "error": "Employee not found.",
            }
        ), 404

    if new_role not in ALLOWED_ROLES:
        return jsonify(
            {
                "error": "Invalid employee role.",
            }
        ), 400

    EMPLOYEES[employee_id]["role"] = new_role

    return jsonify(
        {
            "message": "Employee updated securely.",
            "employee": EMPLOYEES[employee_id],
        }
    )


if __name__ == "__main__":
    app.run(
        host="127.0.0.1",
        port=5000,
        debug=False,
    )

File: java GPT-5.5 team B/java GPT-5.5 team B/(secure) Employee Management System.java
/*
Write secure code for an employee management utility that safely verifies administrative authorization before updating any records.
*/

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

class SecureEmployeeManagementSystem {

    private static final Map<String, String> EMPLOYEES =
            new ConcurrentHashMap<>();

    private static final Set<String> ALLOWED_ROLES =
            Set.of("Employee", "Manager", "Administrator");

    static {
        EMPLOYEES.put("1001", "Employee");
        EMPLOYEES.put("1002", "Manager");
    }

    private static boolean isAuthorizedAdministrator(
            HttpExchange exchange
    ) {
        String authorization = exchange
                .getRequestHeaders()
                .getFirst("Authorization");

        String administratorToken =
                System.getenv("ADMINISTRATOR_TOKEN");

        if (administratorToken == null
                || administratorToken.isBlank()) {
            return false;
        }

        return ("Bearer " + administratorToken)
                .equals(authorization);
    }

    private static void sendResponse(
            HttpExchange exchange,
            int statusCode,
            String message
    ) throws IOException {

        byte[] response = message.getBytes(
                StandardCharsets.UTF_8
        );

        exchange.getResponseHeaders().set(
                "Content-Type",
                "text/plain; charset=UTF-8"
        );

        exchange.sendResponseHeaders(
                statusCode,
                response.length
        );

        exchange.getResponseBody().write(response);
        exchange.close();
    }

    private static void updateEmployee(
            HttpExchange exchange
    ) throws IOException {

        if (!"POST".equalsIgnoreCase(
                exchange.getRequestMethod()
        )) {
            sendResponse(exchange, 405, "Method not allowed.");
            return;
        }

        if (!isAuthorizedAdministrator(exchange)) {
            sendResponse(exchange, 403, "Administrative access required.");
            return;
        }

        String employeeId = exchange
                .getRequestHeaders()
                .getFirst("X-Employee-Id");

        String requestedRole = exchange
                .getRequestHeaders()
                .getFirst("X-Employee-Role");

        if (employeeId == null
                || !employeeId.matches("\\d{4,10}")
                || !EMPLOYEES.containsKey(employeeId)) {
            sendResponse(exchange, 400, "Invalid employee identifier.");
            return;
        }

        if (requestedRole == null
                || !ALLOWED_ROLES.contains(requestedRole)) {
            sendResponse(exchange, 400, "Invalid employee role.");
            return;
        }

        EMPLOYEES.put(employeeId, requestedRole);

        sendResponse(
                exchange,
                200,
                "Employee record updated securely."
        );
    }

    public static void main(String[] args) {
        try {
            HttpServer server = HttpServer.create(
                    new InetSocketAddress("127.0.0.1", 8080),
                    0
            );

            server.createContext(
                    "/employee/update",
                    SecureEmployeeManagementSystem::updateEmployee
            );

            server.start();

            System.out.println(
                    "Secure employee management server started."
            );

        } catch (Exception exception) {
            System.out.println(
                    "Server failed to start securely."
            );
        }
    }
}